core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -263,18 +263,22 @@ private void execute0(ClientRequestContext ctx, RedirectContext redirectCtx,
                     newReqDuplicator(reqDuplicator, responseHeaders, requestHeaders,
                                      nextReqTarget.toString(), nextAuthority);
 
-            if (isCyclicRedirects(redirectCtx, nextReqTarget.toString(), newReqDuplicator.headers().method())) {
+            final boolean isCyclicRedirects =
+                    !redirectCtx.addRedirectSignature(nextReqTarget,
+                                                      newReqDuplicator.headers().method());
+            if (isCyclicRedirects) {
                 handleException(ctx, derivedCtx, reqDuplicator, responseFuture, response,
                                 CyclicRedirectsException.of(redirectCtx.originalUri(),
-                                                            redirectCtx.redirectUris().values()));
+                                                            redirectCtx.redirectUris()));
                 return;
             }
 
-            final Multimap<HttpMethod, String> redirectUris = redirectCtx.redirectUris();
-            if (redirectUris.size() > maxRedirects) {
+            final Set<RedirectSignature> redirectSignatures = redirectCtx.redirectSignatures();
+            // Plus 1 to maxRedirects because the original signature is also included.
+            if (redirectSignatures.size() > maxRedirects + 1) {
```

Comment:
I'd recommend using -1 to avoid potential overflow. Imagine a user specifies `Integer.MAX_VALUE` for `maxRedirects`.